accounting/src/delta/delta_data_collection/mod.rs

```diff
@@ -14,52 +14,133 @@
 // limitations under the License.
 
 pub mod undo;
+use self::undo::*;
 
-use std::collections::BTreeMap;
+use std::collections::{btree_map::Entry, BTreeMap};
 
 use serialization::{Decode, Encode};
 
 use crate::error::Error;
 
-use self::undo::*;
+///
+/// Basic primitive that represents a difference introduced to a data.
+///
+/// Following properties are defined for deltas:
+/// - Associativity:
+///   (d1 + d2) + d3 = d1 + (d2 + d3), where '+' is the combine operation implemented by `combine_delta_data`
+///
+///   Associativity property doesn't hold if a combine operation produces an error, e.g.:
+///   (Delta(Some, None) + Delta(Some, None)) + Delta(None, Some) = Error + Delta(None, Some)
+///   Delta(Some, None) + (Delta(Some, None) + Delta(None, Some)) = Delta(Some, None) + Delta(Some, Some) = Error
+///
+/// - Inversion in the sense that:
+///   (d + d.invert()) + d = d
+///
+#[derive(PartialEq, Eq, Clone, Encode, Decode, Debug)]
+pub struct DataDelta<T> {
+    prev: Option<T>,
+    next: Option<T>,
```

Comment:
Naming these `prev` and `next` is kind of confusing... I suggest something `cur` and `new`. Or `from` and `to`. The labels prev/next usually indicate iteration.